File: examples/eloquent/app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;

/**
 * The primary key is not called `id` and has a column of its own. An order cannot be removed
 * while it has line items (Restrict), and its account cannot be removed while it has orders
 * (NoAction). The only timestamp is `@updatedAt`, under a name of its own.
 */
class Order extends Model
{
    use PrismaDates;

    const CREATED_AT = null;
    const UPDATED_AT = 'changed_at';

    protected $table = 'orders';

    protected $primaryKey = 'order_number';

    protected $fillable = [
        'account_id',
        'status',
        'note',
        'placed_at',
        'public_id',
        'gift_for',
    ];

    protected $attributes = [
        'status' => 'PENDING',
    ];

    protected $casts = [
        'account_id' => 'integer',
        'status' => OrderStatus::class,
        'placed_at' => 'datetime',
    ];

    public function save(array $options = [])
    {
        if (! $this->exists) {
            if (! array_key_exists('public_id', $this->attributes)) {
                $this->setAttribute('public_id', (string) Str::ulid());
            }
            if (! array_key_exists('placed_at', $this->attributes)) {
                $this->setAttribute('placed_at', $this->freshTimestamp());
            }
        }

        return parent::save($options);
    }

    public function account(): BelongsTo
    {
        return $this->belongsTo(Account::class, 'account_id');
    }

    public function giftTarget(): BelongsTo
    {
        return $this->belongsTo(Account::class, 'gift_for', 'handle');
    }

    public function lineItems(): HasMany
    {
        return $this->hasMany(LineItem::class, 'order_number');
    }
}

File: examples/ecto/lib/shop/order.ex
defmodule Shop.Order do
  use Ecto.Schema
  @moduledoc """
  The primary key is not called `id` and has a column of its own: Ecto calls it `:id` with a
  `source:`. An order cannot be removed while it has line items (Restrict), and its account
  cannot be removed while it has orders (NoAction, which SQLite checks at the end of the
  statement).
  """

  @primary_key {:id, :id, autogenerate: true, source: :order_number}
  @timestamps_opts [type: Shop.PrismaDateTime, autogenerate: {Shop.PrismaDateTime, :autogenerate, []}]

  @type t :: %__MODULE__{
          id: integer(),
          status: atom(),
          note: String.t() | nil,
          placed_at: DateTime.t(),
          updated_at: DateTime.t(),
          synced_at: DateTime.t(),
          account: Shop.Account.t() | nil,
          gift_target: Shop.Account.t() | nil,
          line_items: [Shop.LineItem.t()]
        }

  schema "orders" do
    field(:status, Ecto.Enum, values: [:PENDING, :PAID, :SHIPPED, :CANCELLED], default: :PENDING)
    field(:note, :string)
    field(:placed_at, Shop.PrismaDateTime, autogenerate: true)
    belongs_to(:account, Shop.Account, foreign_key: :account_id)
    belongs_to(:gift_target, Shop.Account, foreign_key: :gift_for, type: :string, references: :handle)
    has_many(:line_items, Shop.LineItem, foreign_key: :order_number)
    timestamps(inserted_at: false, updated_at_source: :changed_at)
    timestamps(inserted_at: false, updated_at: :synced_at)
  end
end